### PR TITLE
Manual merge of 3.0.x into master

### PR DESCRIPTION
**Pull request by greg0ire**
Blocked by https://github.com/phpstan/phpstan/issues/3335

### CODE REVIEW

**Comment by morozov** [comment]
Why do we need to merge two changes at once?

**Comment by greg0ire** [comment]
What do you mean by "2 changes"?

**Comment by greg0ire** [comment]
@morozov only the 7 last commits are new and need to be reviewed

**Comment by greg0ire** [comment]
@morozov it's complicated to do the merge up with your changes, I'm not super sure. Can you please do the merge up of 74eca6ba7 ? That way I only have my part to take care about…

**Comment by morozov** [comment]
@greg0ire not sure I understand the ask. 74eca6b is currently part of `3.0.x`. Are you asking me to merge `3.0.x` into `master`? Looks like this is what you're doing in this PR.

As for the change itself, it looks trivial: the method is moved from one interface to another and is implemented in the class which now needs to implement it.

Please let me know how exactly I can help you.

**Comment by greg0ire** [comment]
> @greg0ire not sure I understand the ask. 74eca6b is currently part of 3.0.x. Are you asking me to merge 3.0.x into master? Looks like this is what you're doing in this PR.

No I'm just asking you to merge 74eca6b into master. You don't have to handle the merge for my Psalm related changes, but only for your changes.

> As for the change itself, it looks trivial: the method is moved from one interface to another and is implemented in the class which now needs to implement it. The same change already exists in master where it was backported from.

> Please let me know how exactly I can help you.

I'll give it another try since it does look trivial, but 74eca6b might have ancestors that are not merged either, not sure if the conflict was about them or 74eca6b

**Comment by greg0ire** [comment]
Gave it another try, and I'm still uncomfortable with this. Please try the following on master:
```
git merge 74eca6ba7fdd6a48af0c450a486d1f77f950a041
git push fork
* review*
git push origin
```

**Comment by morozov** [comment]
@greg0ire please see https://github.com/doctrine/dbal/pull/4038.

> As you see, it's far from being just that one trivial commit

Yeah, its the fetch mode rework and the corresponding deprecation backports.

**Comment by greg0ire** [comment]
Thank you!!!

**Comment by greg0ire** [comment]
@morozov ready for review (the last 8 commits are only on this branch)

File: phpcs.xml.dist [hunk 1]
```diff
@@ -113,7 +113,7 @@
 
     <!-- The SQLSRV_* functions are defined in the upper case by the sqlsrv extension and violate the standard
          see https://docs.microsoft.com/en-us/sql/connect/php/constants-microsoft-drivers-for-php-for-sql-server -->
-    <rule ref="Squiz.PHP.LowercasePHPFunctions.UseStatementUppercase">
-        <exclude-pattern>src/Driver/SQLSrv/SQLSrvStatement.php</exclude-pattern>
+    <rule ref="Squiz.PHP.LowercasePHPFunctions">
+        <exclude-pattern>lib/Doctrine/DBAL/Driver/SQLSrv/SQLSrvStatement.php</exclude-pattern>
```
Comment (morozov) [inline phpcs.xml.dist:117]: Looks like this was incorrectly resolved.

File: tests/Platforms/AbstractPostgreSQLPlatformTestCase.php [hunk 2]
```diff
@@ -19,9 +21,12 @@
 
 abstract class AbstractPostgreSQLPlatformTestCase extends AbstractPlatformTestCase
 {
-    /** @var PostgreSQL94Platform */
+    /** @var PostgreSQL94Platform|PostgreSQL100Platform */
```
Comment (morozov) [inline tests/Platforms/AbstractPostgreSQLPlatformTestCase.php:24]: Why is this needed? `PostgreSQL100Platform` is a subtype of `PostgreSQL94Platform`.

Comment (greg0ire) [reply tests/Platforms/AbstractPostgreSQLPlatformTestCase.php:24]: Not needed, I just didn't notice that was the case.

**Comment by greg0ire** [comment]
@morozov can we merge as is? I handled your review in https://github.com/doctrine/dbal/pull/4040, since it was already wrong on 3.0.x

**Review by morozov — DISMISSED** [review]

**Comment by greg0ire** [comment]
@morozov sorry, please approve again, the last push contained the fixes that were intended for #4040 and missed the last commit without which I don't understand how it could pass…

**Review by morozov — APPROVED** [review]
